### ISSUE TITLE
[BUG] ZeRO++ hpZ on llama2-7b gets zero loss

### ISSUE THREAD

**Issue by yundai424** [microsoft/DeepSpeed#5059]
**Describe the bug**
With DeepSpeed built from source on latest master branch (i.e. https://github.com/microsoft/DeepSpeed/pull/4906 is in), hpZ llama2-7b and the training loss will appear 0 starting the 3rd training step:
```
{'loss': 2.1364, 'learning_rate': 0.0, 'epoch': 0.0, 'max_steps': 50, 'global_step': 1, 'current_step_time_seconds': 19.497754335403442, 'average_step_time_seconds': 19.497755765914917, 'estimated_time_to_completion_seconds': 955.3900325298309, 'estimated_total_training_time_seconds': 974.8877882957458}
{'loss': 2.0157, 'learning_rate': 2e-05, 'epoch': 0.0, 'max_steps': 50, 'global_step': 2, 'current_step_time_seconds': 24.733116388320923, 'average_step_time_seconds': 22.11543607711792, 'estimated_time_to_completion_seconds': 1061.5409317016602, 'estimated_total_training_time_seconds': 1105.771803855896}
{'loss': 0.0, 'learning_rate': 2e-05, 'epoch': 0.0, 'max_steps': 50, 'global_step': 3, 'current_step_time_seconds': 24.691705465316772, 'average_step_time_seconds': 22.974192539850872, 'estimated_time_to_completion_seconds': 1079.787049372991, 'estimated_total_training_time_seconds': 1148.7096269925437}
{'loss': 0.0, 'learning_rate': 2e-05, 'epoch': 0.0, 'max_steps': 50, 'global_step': 4, 'current_step_time_seconds': 25.519466638565063, 'average_step_time_seconds': 23.61051106452942, 'estimated_time_to_completion_seconds': 1086.0835089683533, 'estimated_total_training_time_seconds': 1180.525553226471}
```

This however cannot be reproduced on GPT2. 

**To Reproduce**
1. Training script:
```python
import transformers
import datasets
import trl
from dataclasses import dataclass, field
import peft
import torch
import time
import callbacks

MODEL_NAME = "meta-llama/Llama-2-7b"

def formatting_func(example):
    output_texts = []
    for i in range(len(example['question'])):
        choices = ""
        for j in range(len(example["choices"][i])):
            choices += f"{j+1}. {example['choices'][i][j]}\n"
        output_texts.append(f"""Below is a question and subject and multiple choice answers, choices separated by a comma. Please select the best answer for the question.

### Question:
{example["question"][i]}

### Subject:
{example["subject"][i]}

### Choices:
{choices}

### Answer:
{example["answer"][i]}"""
)
    return output_texts



def main():
    parser = transformers.HfArgumentParser(transformers.TrainingArguments)
    training_args = parser.parse_args_into_dataclasses()
    dataset = datasets.load_dataset('cais/mmlu','all')
    dataset_train, dataset_eval = dataset['auxiliary_train'], dataset['test']

    config = transformers.AutoConfig.from_pretrained(MODEL_NAME)
    config.num_hidden_layers = 9
    model = transformers.AutoModelForCausalLM.from_config(config)

    tokenizer = transformers.AutoTokenizer.from_pretrained(MODEL_NAME)
    tokenizer.pad_token = tokenizer.eos_token

    trainer = trl.SFTTrainer(
        model=model,
        tokenizer=tokenizer,
        train_dataset=dataset_train,
        eval_dataset=dataset_eval,
        formatting_func=formatting_func,
        max_seq_length=512,
        args=training_args
    )
    trainer.train()

if __name__ == "__main__":
    main()
```

2. DS config:
```json
{
    "bf16": {
      "enabled": "auto"
    },
    "optimizer": {
        "type": "AdamW",
        "params": {
            "lr": "auto",
            "betas": "auto",
            "eps": "auto",
            "weight_decay": "auto"
        }
    },

    "scheduler": {
        "type": "WarmupLR",
        "params": {
            "warmup_min_lr": "auto",
            "warmup_max_lr": "auto",
            "warmup_num_steps": "auto"
        }
    },

    "zero_optimization": {
        "stage": 3,
        "overlap_comm": true,
        "contiguous_gradients": true,
        "stage3_gather_16bit_weights_on_model_save": true,
        "zero_hpz_partition_size": 4
    },

    "gradient_accumulation_steps": "auto",
    "gradient_clipping": "auto",
    "train_batch_size": "auto",
    "train_micro_batch_size_per_gpu": "auto",
}
```

3. launch script:
```
export LOCAL_WORLD_SIZE=4
export NUM_NODES=2
export WORLD_SIZE=$((LOCAL_WORLD_SIZE * NUM_NODES))
export NCCL_DEBUG=INFO

torchrun --nnodes=$NUM_NODES --nproc-per-node=$LOCAL_WORLD_SIZE \
      --rdzv_endpoint="$MASTER_ADDR:$MASTER_PORT" --rdzv_id=1234 --rdzv_backend=c10d \
      src/train.py \
      --bf16 True \
      --max_steps 50 \
      --per_device_train_batch_size 2 \
      --per_device_eval_batch_size 2 \
      --save_strategy "no" \
      --learning_rate 2e-5 \
      --logging_steps 1 \
      --report_to none \
      --output_dir "$OUTPUT_FULL_PATH" \
      --deepspeed zero3_hpz.json
```

This is tested on llama2-7b-hf using MMLU dataset.

**Expected behavior**
convergence should be the same as that without hpZ

**ds_report output**
```
--------------------------------------------------
DeepSpeed C++/CUDA extension op report
--------------------------------------------------
NOTE: Ops not installed will be just-in-time (JIT) compiled at
      runtime if needed. Op compatibility means that your system
      meet the required dependencies to JIT install the op.
--------------------------------------------------
JIT compiled ops requires ninja
ninja .................. [OKAY]
--------------------------------------------------
op name ................ installed .. compatible
--------------------------------------------------
 [WARNING]  async_io requires the dev libaio .so object and headers but these were not found.
 [WARNING]  async_io: please install the libaio-devel package with yum
 [WARNING]  If libaio is already installed (perhaps from source), try setting the CFLAGS and LDFLAGS environment variables to where it can be found.
async_io ............... [NO] ....... [NO]
fused_adam ............. [NO] ....... [OKAY]
cpu_adam ............... [NO] ....... [OKAY]
cpu_adagrad ............ [NO] ....... [OKAY]
cpu_lion ............... [NO] ....... [OKAY]
 [WARNING]  Please specify the CUTLASS repo directory as environment variable $CUTLASS_PATH
evoformer_attn ......... [NO] ....... [NO]
fused_lamb ............. [NO] ....... [OKAY]
fused_lion ............. [NO] ....... [OKAY]
inference_core_ops ..... [NO] ....... [OKAY]
cutlass_ops ............ [NO] ....... [OKAY]
quantizer .............. [NO] ....... [OKAY]
ragged_device_ops ...... [NO] ....... [OKAY]
ragged_ops ............. [NO] ....... [OKAY]
random_ltd ............. [NO] ....... [OKAY]
 [WARNING]  sparse_attn requires a torch version >= 1.5 and < 2.0 but detected 2.0
 [WARNING]  using untested triton version (2.0.0), only 1.0.0 is known to be compatible
sparse_attn ............ [NO] ....... [NO]
spatial_inference ...... [NO] ....... [OKAY]
transformer ............ [NO] ....... [OKAY]
stochastic_transformer . [NO] ....... [OKAY]
transformer_inference .. [NO] ....... [OKAY]
--------------------------------------------------
DeepSpeed general environment info:
torch install path ............... ['/home/jobuser/.local/lib/python3.10/site-packages/torch']
torch version .................... 2.0.0+cu118
deepspeed install path ........... ['/home/jobuser/.local/lib/python3.10/site-packages/deepspeed']
deepspeed info ................... 0.13.2+feffed1, feffed1, master
torch cuda version ............... 11.8
torch hip version ................ None
nvcc version ..................... 11.8
deepspeed wheel compiled w. ...... torch 2.0, c
```

**System info (please complete the following information):**
 - OS: CentOS7
 - GPU count and types: 2 nodes each with 8 A100s (only enabled 4 each)
 - Interconnects: no we only get ethernet for inter-node comm
 - Python version: 3.10

**Launcher context**
`torchrun` launched on 2 nodes separatedly

**Additional context**
This can be reproduced with `transformers==4.37.2`

**Docker context**
unfortunately no but it's a simple one w/ cu118 and py310 

Will update this as soon as I have a more concrete finding.

**Comment by yundai424** [comment]
Update: was trying to reproduce with a more lightweight example using [this test case](https://github.com/microsoft/DeepSpeed/blob/master/tests/unit/runtime/zero/test_zeropp.py#L192) added by @HeyangQin by simply adding `model.gradient_checkpointing_enable()` after loading model with `from_pretrained` but the loss is consistent between zero3 and zeropp, no hanging observed either (gist [here](https://gist.github.com/yundai424/560bd5acc7323a3ae00c0145ef7da33b)) 😓

**Comment by yundai424** [comment]
was trying to reproduce with a much simplified training loop without HF trainer but we can't observe the nan loss problem:
```
import deepspeed.comm as dist
import deepspeed
from deepspeed.runtime.zero.config import DeepSpeedZeroConfig
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer
from torch.utils.data import DataLoader
import numpy as np
import pytest
import transformers


MODEL_NAME = "/shared/public/models/llama-2-7b-hf"
DS_CONFIG = {
    'bf16': {'enabled': True}, 
    'optimizer': {
        'type': 'AdamW', 
        'params': {
            'lr': 2e-05, 
            'betas': [0.9, 0.999], 
            'eps': 1e-08, 
            'weight_decay': 0.0
        }
    }, 
    'scheduler': {
        'type': 'WarmupLR', 
        'params': {
            'warmup_min_lr': 0, 
            'warmup_max_lr': 2e-05, 
            'warmup_num_steps': 0
        }
    }, 
    'zero_optimization': {
        'stage': 3, 
        'overlap_comm': True, 
        'contiguous_gradients': True, 
        'sub_group_size': 1e9, 
        'reduce_bucket_size': 4096*4096, 
        'stage3_prefetch_bucket_size': 4096*4096*0.9, 
        'stage3_max_live_parameters': 1e9, 
        'stage3_max_reuse_distance': 1e9, 
        'stage3_gather_16bit_weights_on_model_save': True, 
        'zero_hpz_partition_size': 4
    },
    'gradient_accumulation_steps': 1, 
    'gradient_clipping': 1.0, 
    'steps_per_print': float('inf'), 
    'train_batch_size': 16, 
    'train_micro_batch_size_per_gpu': 2, 
    'wall_clock_breakdown': False, 
}

def load_and_prepare_data(model_name):
    """Load model, tokenizer and dataset, and prepare data loader."""
    from datasets import load_from_disk

    # Load model and tokenizer
    config = transformers.AutoConfig.from_pretrained(model_name)
    config.num_hidden_layers = 9
    model = transformers.AutoModelForCausalLM.from_config(config).to(torch.bfloat16)
    tokenizer = AutoTokenizer.from_pretrained(model_name)
    tokenizer.pad_token = tokenizer.eos_token

    # Load and tokenize dataset
    dataset = load_from_disk("/shared/public/data/yudai/wikitext_sample")

    def tokenize_function(examples):
        # Tokenize and ensure 'labels' are the same as 'input_ids'
        tokenized_output = tokenizer(examples["text"], padding="longest", truncation=True, return_tensors='pt', max_length=256)
        tokenized_output["labels"] = tokenized_output["input_ids"].clone()
        return tokenized_output

    tokenized_dataset = dataset.map(tokenize_function, batched=True).filter(lambda x: x['text'])
    tokenized_dataset.set_format('torch', columns=['input_ids', 'attention_mask', 'labels'])

    # Create data loader
    data_loader = DataLoader(tokenized_dataset, batch_size=2, shuffle=False)
    return model, data_loader

def get_loss(model, data_loader, config_dict, step=500):
    """Train the model and calculate average loss."""
    # Initialize DeepSpeed
    model, _, _, _ = deepspeed.initialize(model=model, model_parameters=model.parameters(), config=config_dict, dist_init_required=True)
    dist.barrier()
    model.train()

    # Training loop
    losses = []
    for n, batch in enumerate(data_loader):
        if n >= step:
            break
        batch = {k: v.to(model.device) for k, v in batch.items()}
        outputs = model(**batch)
        loss = outputs.loss
        if torch.distributed.get_rank() == 0:
            print(f"loss: {loss}")
        model.backward(loss)
        model.step()
        losses.append(loss.item())

    return np.nanmean(losses[-100:])

if __name__ == "__main__":
    torch.manual_seed(0)
    model, data_loader = load_and_prepare_data(model_name)
    zeropp_loss = get_loss(model, data_loader, DS_CONFIG)
```

In the above script I explicitly set DS config to be the same as what HF/accelerate automatically generates.

**Comment by yundai424** [comment]
Some more update:
I recorded the `ds_id` of the param that behaves differently, and added logging [here](https://github.com/microsoft/DeepSpeed/blob/9d2660d2a3fac767972f01ac96858b2605ffc0e4/deepspeed/runtime/zero/partition_parameters.py#L1169) for this param and a "normal" param every time they're enqueued for allgather. It appears that the second allgather on this param within each step will take place before backprop (while the 2nd allgather on other params are, as expected, after backward starts):

```python
        def all_gather_coalesced(params: Iterable[Parameter],
                                 forward: bool = True,
                                 safe_mode: bool = False,
                                 quantize: bool = False) -> AllGatherCoalescedHandle:

            # fetches from nvme if the partition is not available and in nvme
            self._ensure_availability_of_partitioned_params(params)

            if self.num_partitions == 1:
                return _no_gather_coalesced(params)

            for param in params:
                if param.ds_status != ZeroParamStatus.NOT_AVAILABLE:
                    raise RuntimeError(param.ds_summary())
                param.ds_status = ZeroParamStatus.INFLIGHT
                if torch.distributed.get_rank() == 0:
                    if param.ds_id == 39:
                        print("this param")
                    elif param.ds_id == 2:
                        print(param.ds_id)
```

```
---starting forward---
2
[2024-02-06 18:36:21,904] [WARNING] [parameter_offload.py:87:_apply_to_tensors_only] A module has unknown inputs or outputs type (<class 'transformers.cache_utils.DynamicCache'>) and the tensors embedded in it cannot be detected. The ZeRO-3 hooks designed to trigger before or after backward pass of the module relies on knowing the input and output tensors and therefore may not get triggered properly.
10
this param
---finish forward---
this param
10
2
---finish backward and step---
{'loss': 11.2538, 'learning_rate': 0.0, 'epoch': 0.0}                                                   
  2%|█▎                                                                  | 1/50 [00:04<03:34,  4.37s/it]starting forward
2
10
this param
this param
---finish forward---
10
2
---finish backward and step---
{'loss': 10.9484, 'learning_rate': 2e-05, 'epoch': 0.0}                                                 
  4%|██▋                                                                 | 2/50 [00:07<02:43,  3.41s/it]starting forward
2
10
this param
this param
---finish forward---
NaN detected in module.model.layers.4.self_attn.v_proj.weight during after forward, param stage ZeroParamStatus.INFLIGHT, param ds id 39
10
2
---finish backward and step---
{'loss': 10.9312, 'learning_rate': 2e-05, 'epoch': 0.0}                                                 
  6%|████                                                                | 3/50 [00:09<02:27,  3.14s/it]
```

**Comment by ighoshsubho** [comment]
I would like to work on this.

**Comment by ByronHsu** [comment]
1. The issue can be reproduced even if we initialize the 2nd tensor as torch.rand, which seems to break our hypothesis.
2. Add a check for the 2nd tensor before all gather. And we can observe that `param_ds_tensor` does contain nan.
```python
if torch.isnan(param_ds_tensor).any().item():
    print("param ds tensor contains nan!!!")
    exit()
    
handles = _dist_allgather_fn(
    param_ds_tensor.to(get_accelerator().current_device_name()),
    param_buffer,
    ds_process_group,
)
```
```bash
param ds tensor contains nan!!!
```

3. Add a nan check in the if condition again to see if it could be due to async problem

```python
if torch.isnan(param_ds_tensor).any().item():
    import time
    time.sleep(2)
    print(torch.isnan(param_ds_tensor).any().item())
    print("param ds tensor contains nan!!!")
    exit()
```

```
param ds tensor contains nan!!!
False
```

It means that the ds tensor contains nan in the beginning, but after 2 seconds, the nan is gone!! 

4. Remove the above two check and add a check after 2nd partitioning. In this case, the nan doesn’t happen (what???)

```python
if torch.isnan(param.ds_secondary_tensor).any().item():
 print("param ds tensor contains nan after partition!!!")
 exit()
print_rank_0(f"{param.ds_id} partitioned type {param.dtype} dev {param.device} shape {param.shape}",
          force=False)
```

5. `copy_` is non-blocking if it is d2d 

![image](https://github.com/microsoft/DeepSpeed/assets/24364830/14bb4362-91b5-44dc-be83-3fe73d3c7261)

6. Instead of injecting nan check in deepspeed code, i moved the check to huggingface trainer. I added checks in multiple places, including “before forward”, “after forward”, “before backward”, “after backward”, “before step”, and “after step”. I observed that nan happens at “after backward”, which means during backward, the weights get mutated to NaN, thus causing NaN in grad. Technically, backward shouldn't change weights at all. It is likely that the collective ops messed up the weights.

7. With torch.cuda.synchronize() added before all gather, if we tune stage3_param_persistence_threshold to zero, we can still observe nan issue.

8. NaN issue is not reproducible with a ~7B linear model

**Comment by ByronHsu** [comment]
Some questions that bother us and our hypothesis


Question: Why adding the check “after 2nd partitioning” can prevent nan issue, but adding the check “before all gather” still makes the issue present?

Hypothesis:
“torch.isnan(tensor).any().item()” essentially copies the data from device to host thus serves as a sync point. However, “After 2nd partitioning” is running on the forward thread but “Before all gathering” is running on the backward thread. If we put the check at the same thread where the copying happens, it will sync due to .item(). On the other hand, if we put the check at a different thread, .item() does not wait. 

Question: Why initializing 2nd tensors with torch.randn doesn’t fix the issue? 

Hypothesis 1:

[copying                ]
                    [all gather        ]
Even if we initialize 2nd tensors as non-nan, if we probe the values of the tensor during copying, we will get some nan values.

Hypothesis 2:
The all gather doesn’t get any nan, but instead it gets a randn value, and then uses the value for optimizer step, causing nan in the updated weights.

**Comment by czczup** [comment]
I meet the same problem, my loss is zero when I use zero++

**Comment by yundai424** [comment]
@czczup the fix has been out in the latest release. Could you try use 0.13.5 and see if it still happens?
